### ISSUE TITLE
Reorganization and name change

### ISSUE THREAD

**Issue by stuarteberg** [ilastik/ilastik-publish-packages#18]
I'm currently using this script to build FlyEM's package stack, and I'd like to start recommending it to others when appropriate.  What do you think of the following proposal:

- Reorganize the files into a proper python package with `setup.py`, `entry_points`, a conda recipe, docs, etc.  Heck, maybe even some tests...
- Remove project-specific configuration files like `ilastik-recipe-specs.yaml`, etc.
- Rename the package to something more generic.  Any opinions here?  Maybe something like, I don't know, `build-recipes`, or `publish-recipes`, or `publish-stack`, etc.

**Comment by stuarteberg** [comment]
>When renaming it, we should keep in mind what the actual use-case for this is. 

Okay, how about `publish-conda-stack`?


>we build everything in order ... Do we want to keep it that way ... Or somehow solve the dependency graph for all the packages specified and build them in order

I think changing this behavior is beyond the scope of what I have time for in the near term.  In the long term, I'm still a little wary of this.  Of course, Carsten and I originally considered performing an automatic toposort as you propose, but there are some potential pitfalls:

- Circular dependencies are a potential issue.  OK, in practice, there is usually only one circular dependency in our stack, and it is handled in conda by [this `if` statement][1] I wrote a [long time ago][2], so if we use their `toposort()` function, it should be okay.

- If I'm not mistaken, then performing a provably-correct toposort requires obtaining the ENTIRE dependency graph for all packages in your stack -- even ones that aren't listed in your `recipe-specs.yaml` file.  Running `conda search` on so many packages takes time.  Maybe there is a better way to fetch the dependencies for lots of packages -- after all, the conda solver manages it -- but it would take some investigation.

   It is tempting to just download the dependency info for the particular packages you're building, but that might not be sufficient in some cases.  Perhaps such cases are a bit contrived, but it's at least theoretically possible.  If a stack has dependencies A <- B <- C, but you only rebuild A and C, `toposort()` might emit A and C in an arbitrary order (unless you include B in the input).

   OK, I think I've come up with a concrete example, if you care to indulge me.  :-) Suppose my stack consisted of `dvid` <-- `pydvid`, <-- `ilastik`.  And suppose `dvid` has a REST API to return the server name as JSON: `{"server": "foobar"}`, and `ilastik` reads the message via this line:

   ```
   server_name = pydvid.fetch_server_info()["server"]
   ```

   So far, so good.  But now suppose DVID changes its API.  Now it returns `{"server-NAME": "foobar"}`.  Since `pydvid` was just a pass-through, we wouldn't rebuild `pydvid`.  But both `dvid` and `ilastik` would need to be rebuilt.  Hence, this is the A/C scenario I mentioned above.  Such cases would be rare, but potentially confusing if they do occur.

Maybe a "best-of-both-worlds" solution would be to perform a check based on the packages the user is building, and emit an error if they appear NOT to be toposorted in the `yaml` file.  So, users are still required to toposort the inputs manually, but we'll help them spot mistakes.

[1]: https://github.com/conda/conda/blob/5ef35ba895c0b577faf0e6f1f4c4810d7d8d4eec/conda/common/toposort.py#L101-L109

[2]: https://github.com/conda/conda/pull/1614

**Comment by k-dominik** [comment]
Thank you for those very concise thoughts.

> Maybe a "best-of-both-worlds" solution would be to perform a check based on the packages the user is building, and emit an error if they appear NOT to be toposorted in the yaml file. So, users are still required to toposort the inputs manually, but we'll help them spot mistakes.

Maybe this is the way to go! It's great that this functionality exists, already. And failure with some nice hint on where to spot it is already very nice to have.

> OK, I think I've come up with a concrete example, if you care to indulge me. :-) Suppose my stack consisted of dvid <-- pydvid, <-- ilastik. And suppose dvid has a REST API to return the server name as JSON: {"server": "foobar"}, and ilastik reads the message via this line:

But, then, where does the `pydvid` package come from? Isn't this kind of pathological, already? I know, `pydvid` might be `noarch` and sourced from a different channel, but depends on some host specific package that one rebuilds for some different `OS` or something?

I just wanted to raise the point, that one might end up with a stack that doesn't necessarily reflect the intentions, like, building `vigra` against a specific version of `hdf5` and not having the order right. This would be unnoticed (build might still pass, `hdf5` is available somewhere else)...

**Comment by stuarteberg** [comment]
>Thank you for those very concise thoughts.

LOL, sorry.  You know by now that brevity difficult concept for me.   :-P

>Maybe this is the way to go!

OK, cool.  Even so, I think I won't tackle this just yet.

>But, then, where does the pydvid package come from?

I'm speculating that it's one of our own packages, but that it required no rebuild (none of its source code changed), so the user typed `./build-recipes.py ilastik-recipe-specs.py ilastik dvid`

>Isn't this kind of pathological, already?

Yes, perhaps I'm worrying too much over unrealistic scenarios.

**Comment by k-dominik** [comment]
Okay, so we'll keep it "linear", as is, in the beginning and just make the current version proper. Any idea on where to put this then? I guess the question is, whether to have it here in the ilastik org or at flyem? I'm fine either way. Do you have any time frame in mind, do you want to do it alone, or wouldn't you mind if I pitched in? I'm sure up for it

**Comment by k-dominik** [comment]
Just a progress update: I didn't have as much time to invest as I hoped and I also had to work on a cookiecutter first, to go with the project ;).
In any case, I've pushed the current status here: https://github.com/ilastik/ilastik-publish-packages/tree/rework

I mainly reorganized everything into a package with a proper main and a basic conda recipe. There is still a lot to do. However, when installed, one can already use the script after installing:

Example invocation:

```bash
publish-conda-stack ilastik-recipe-specs
```

I uploaded the current package to `ilastik-forge`. Install:

```
conda install -c ilastik-forge -c conda-forge publish-conda-stack
```
